adl/rot13adl/rot13reification.go

```diff
@@ -0,0 +1,72 @@
+package rot13adl
+
+import (
+	"fmt"
+
+	"github.com/ipld/go-ipld-prime"
+)
+
+// Reify examines data in a Node to see if it matches the shape for valid substrate data for this ADL,
+// and if so, synthesizes and returns the high-level view of the ADL.
+// If it succeeds in recognizing the raw data as this ADL,
+// Reify returns a new Node which exhibits the logical behaviors of the ADL;
+// otherwise, it returns an error.
+//
+// The input data can be any implementation of ipld.Node;
+// it will be considered purely through that interface.
+//
+// If your application is expecting ADL data, this pipeline can be optimized
+// by using the SubstratePrototype right from the start when unmarshalling;
+// then, Reify can detect if the rawRoot parameter is of that implementation,
+// and it can save some processing work internally that can be known to already be done.
+//
+// Reification will generally operate on the data in a single block
+// (e.g. this function will not do any additional block loads and unmarshalling).
+// This is important because some ADLs handle data so large that loading it all
+// eagerly would be impractical (and in some cases outright impossible).
+// However, it also necessarily implies that invalid data may lie beyond
+// one of those lazy loads, and it won't be discovered at the time of Reify.
+//
+// In this demo ADL, we don't have multi-block content at all,
+// so of course we don't have any additional block loads!
+// However, ADL implementations may vary in their approaches to lazy vs eager loading.
+// All ADLs should document their exact semantics regarding this --
+// especially if it has any implications for boundaries of data validity checking.
+//
+// REVIEW: this function is currently not conforming to any particular interface;
+// if we evolve the contract for ADLs to include an interface for reficiation functions,
+// might we need to add context and link loader systems as parameters to it?
+// Not all implementations might need it, as per previous paragraph; but some might.
+// Reification for multiblock ADLs might also need link loader systems as a parameter here
+// so they can capture them as config and hold them for use in future operations that do lazy loading.
```

Comment:
My response here is pretty much the same as my response to Substrate :)

I really, really wish this could be just AssignNode. Maybe we could rethink that API a bit so that we can avoid NewBuilder when it's not needed. Having an extra method documented as "like AssignNode but will save you an alloc in some trivial cases" would be pretty weird.

Reply:
I wish that too, but I think in this case, we're going to end up needing a new function because it's going to be doing a lot more than dodge some allocs.  The need for link loader configuration is going to make for a noticeably different function signature.  It doesn't show up in this demo because the demo isn't doing multiblock stuff, but in a lot of other ADLs, it will.

(Probably.  Unless we figure out some other API for shuffling in those link loader systems from the side, somehow.  But I think you're as up-to-date on the thinking about this as I am, at this point...)